mutable_tree.go

```diff
@@ -1227,48 +1144,73 @@ func (tree *MutableTree) balance(node *Node, orphans *[]*Node) (newSelf *Node, e
 		}
 		if rightBalance <= 0 {
 			// Right Right Case
-			newNode, orphaned, err := tree.rotateLeft(node)
+			newNode, err := tree.rotateLeft(node)
 			if err != nil {
 				return nil, err
 			}
-			*orphans = append(*orphans, orphaned)
 			return newNode, nil
 		}
 		// Right Left Case
-		var rightOrphaned *Node
-
-		right, err := node.getRightNode(tree.ImmutableTree)
-		if err != nil {
-			return nil, err
-		}
-		node.rightHash = nil
-		node.rightNodeKey = 0
-		node.rightNode, rightOrphaned, err = tree.rotateRight(right)
+		node.rightNodeKey = nil
+		node.rightNode, err = tree.rotateRight(rightNode)
 		if err != nil {
 			return nil, err
 		}
-		newNode, leftOrphaned, err := tree.rotateLeft(node)
+		newNode, err := tree.rotateLeft(node)
 		if err != nil {
 			return nil, err
 		}
-
-		*orphans = append(*orphans, right, leftOrphaned, rightOrphaned)
 		return newNode, nil
 	}
 	// Nothing changed
 	return node, nil
 }
 
-func (tree *MutableTree) addOrphans(orphans []*Node) error {
-	for _, node := range orphans {
-		if !node.persisted {
-			// We don't need to orphan nodes that were never persisted.
-			continue
+// getNewNodes gets new created nodes by the changes of the working tree.
+// NOTE: This function clears leftNode/rigthNode recursively and
+// calls _hash() on the given node.
+func (tree *MutableTree) getNewNodes() (map[int32]*Node, error) {
+	newNodes := make(map[int32]*Node)
+	version := tree.version + 1
+
+	nonce := int32(0)
+	var recursiveSpread func(*Node) (*NodeKey, error)
+	recursiveSpread = func(node *Node) (*NodeKey, error) {
+		if node.nodeKey != nil {
+			return node.nodeKey, nil
+		}
+		nonce++
+		node.nodeKey = &NodeKey{
+			version: version,
+			nonce:   nonce,
+		}
```

Comment:
This is a pre-order traversal. As a result, the nonces do not follow the lexicographic order. This leads to increased superfluous merges.


Please try changing to in-order traversal. I would expect the benchmarks to improve with that change

Reply:
you are right, this is why we use the map  and iterate again.